File: services/nuldoc/nuldoc-src/components/Pagination.ts
import { a, div, Element, nav, span } from "../dom.ts";

type Props = {
  currentPage: number;
  totalPages: number;
  basePath: string;
};

export default function Pagination(
  { currentPage, totalPages, basePath }: Props,
): Element {
  if (totalPages <= 1) {
    return div({});
  }

  const pages = generatePageNumbers(currentPage, totalPages);

  return nav(
    { class: "pagination" },
    div(
      { class: "pagination-prev" },
      currentPage > 1
        ? a({ href: pageUrlAt(basePath, currentPage - 1) }, "前へ")
        : null,
    ),
    ...pages.map((page) => {
      if (page === "...") {
        return div({ class: "pagination-elipsis" }, "…");
      } else if (page === currentPage) {
        return div(
          { class: "pagination-page pagination-page-current" },
          span({}, String(page)),
        );
      } else {
        return div(
          { class: "pagination-page" },
          a({ href: pageUrlAt(basePath, page) }, String(page)),
        );
      }
    }),
    div(
      { class: "pagination-next" },
      currentPage < totalPages
        ? a({ href: pageUrlAt(basePath, currentPage + 1) }, "次へ")
        : null,
    ),
  );
}

type PageItem = number | "...";

/**
 * Generates page numbers for pagination display.
 *
 * - Always show the first page
 * - Always show the last page
 * - Always show the current page
 * - Always show the page before and after the current page
 * - If there's only one page gap between displayed pages, fill it
 * - If there are two or more pages gap between displayed pages, show ellipsis
 */
function generatePageNumbers(
  currentPage: number,
  totalPages: number,
): PageItem[] {
  const pages = new Set<number>();
  pages.add(1);
  pages.add(Math.max(1, currentPage - 1));
  pages.add(currentPage);
  pages.add(Math.min(totalPages, currentPage + 1));
  pages.add(totalPages);

  const sorted = Array.from(pages).sort((a, b) => a - b);

  const result: PageItem[] = [];
  for (let i = 0; i < sorted.length; i++) {
    if (i > 0) {
      const gap = sorted[i] - sorted[i - 1];
      if (gap === 2) {
        result.push(sorted[i - 1] + 1);
      } else if (gap > 2) {
        result.push("...");
      }
    }
    result.push(sorted[i]);
  }

  return result;
}

function pageUrlAt(basePath: string, page: number): string {
  return page === 1 ? basePath : `${basePath}${page}/`;
}

File: services/nuldoc/lib/nuldoc/components/pagination.rb
module Nuldoc
  module Components
    class Pagination < DOM::HTMLBuilder
      def initialize(current_page:, total_pages:, base_path:)
        super()
        @current_page = current_page
        @total_pages = total_pages
        @base_path = base_path
      end

      def build
        if @total_pages <= 1
          div
        else
          pages = generate_page_numbers(@current_page, @total_pages)

          nav(class: 'pagination') do
            div(class: 'pagination-prev') do
              a(href: page_url_at(@base_path, @current_page - 1)) { text '前へ' } if @current_page > 1
            end
            pages.each do |page|
              if page == '...'
                div(class: 'pagination-elipsis') { text "\u2026" }
              elsif page == @current_page
                div(class: 'pagination-page pagination-page-current') do
                  span { text page.to_s }
                end
              else
                div(class: 'pagination-page') do
                  a(href: page_url_at(@base_path, page)) { text page.to_s }
                end
              end
            end
            div(class: 'pagination-next') do
              a(href: page_url_at(@base_path, @current_page + 1)) { text '次へ' } if @current_page < @total_pages
            end
          end
        end
      end

      private

      def generate_page_numbers(current_page, total_pages)
        pages = Set.new
        pages.add(1)
        pages.add([1, current_page - 1].max)
        pages.add(current_page)
        pages.add([total_pages, current_page + 1].min)
        pages.add(total_pages)

        sorted = pages.sort

        result = []
        sorted.each_with_index do |page, i|
          if i.positive?
            gap = page - sorted[i - 1]
            if gap == 2
              result.push(sorted[i - 1] + 1)
            elsif gap > 2
              result.push('...')
            end
          end
          result.push(page)
        end

        result
      end

      def page_url_at(base_path, page)
        page == 1 ? base_path : "#{base_path}#{page}/"
      end
    end
  end
end
